Auth: Filter UsedBy results outside of transactions

markylaing opened the pull request:
Additionally, ensure that all UsedBy fields are filtered by what the user can view.

Blocks #12976

markylaing commented:
@tomponline this one should be fine to merge before the driver PR though.

tomponline commented on lxd/api_project.go:
```diff
@@ -195,56 +195,27 @@ func projectsGet(d *Daemon, r *http.Request) response.Response {
 	}
 
-	volumes, err := tx.GetStorageVolumeURIs(ctx, project.Name)
+	entityURLs, err := cluster.GetEntityURLs(ctx, tx.Tx(), project.Name, reportedEntityTypes...)
```
Could we use this pattern for other used by listings perhaps?

markylaing replied:
Most likely yes, but we'll need to write new queries for each entity type that has a "used-by" and each entity type that is displayed in those lists.

tomponline commented on lxd/api_project.go:
```diff
@@ -195,56 +195,27 @@ func projectsGet(d *Daemon, r *http.Request) response.Response {
 	}
 
-	volumes, err := tx.GetStorageVolumeURIs(ctx, project.Name)
+	entityURLs, err := cluster.GetEntityURLs(ctx, tx.Tx(), project.Name, reportedEntityTypes...)
```
This isn't filtering by project though right? So whilst its reducing queries (good) its potentially oukking back lots more rows than needed (bad) (especially instance and storage volume types). For smaller installs fewer queries would be better but at some point the trade off switches into the negative.  Maybe we can add project filter to this function too later.

tomponline replied:
Ah yeah I missed that. Cool.

tomponline commented on test/suites/projects.sh:
```diff
@@ -1,12 +1,12 @@
 test_projects_default() {
   # The default project is used by the default profile
-  lxc project show default | grep -q "/1.0/profiles/default$"
+  lxc project show default | grep -q "/1.0/profiles/default?project=default$"
```
I don't think the api output should be changing here. I don't think we have historically included the default project on the URLs in the used by list.

tomponline reviewed (changes requested):
The used by list shouldn't have default project on its URL

markylaing commented:
> The used by list shouldn't have default project on its URL 

Ok should be an easy fix. Will sort it in the morning.

markylaing commented:
@tomponline I've updated the "used-by" function to drop the project query parameter if the value is the default project. I've also dropped he commit that changed the tests. Should be good to go :)

tomponline commented on lxd/project/permissions.go:
```diff
@@ -1465,22 +1465,71 @@ var aggregateLimitConfigValuePrinters = map[string]func(int64) string{
+	for entityType, urls := range urlsByEntityType {
+		// If only one entry of this type, check directly.
+		if len(urls) == 1 {
+			err := authorizer.CheckPermission(r.Context(), r, urls[0], auth.EntitlementCanView)
```
WRT to

> The given UsedBy URLs may not contain the "project" query parameter.
> Passing them directly to the authorizer is problematic because the OpenFGA datastore depends on this.

This seems like a future bug waiting to happen elsewhere as this is making it every caller's job to rewrite every URL to ensure it always has a project param where appropriate, rather than the authorizer handling it centrally.

Is there a reason you're doing it this way, and can we move the logic into the authorizer?

I wont block the PR on this - but think we should add to the refactoring tasks to make this less brittle going forward.

tomponline replied:
If the authorizer handled it internally then we wouldnt also need to strip the project again later either.

markylaing replied:
Yeah that makes sense. I'll create an issue.
